consumer/consumerconnection/client.go

```diff
@@ -0,0 +1,166 @@
+// # Producers
+//
+// Receivers are typically responsible for obtaining a consumerconnection.Info from the current
+// context and enhancing the consumerconnection.Info with the net.Addr from the peer,
+// storing a new consumerconnection.Info into the context that it passes down. For HTTP
+// requests, the net.Addr is typically the IP address of the client.
+//
+// Components should typically delegate this processing to helpers such
+// as the confighttp or configgrpc packages: both contain interceptors that will
+// enhance the context with the consumerconnection.Info, such that no actions are needed by
+// receivers that are built using confighttp.HTTPServerSettings or
+// configgrpc.GRPCServerSettings.
+//
+// Authenticators are responsible for obtaining a consumerconnection.Info from the current
+// context, enhancing the consumerconnection.Info with an implementation of consumerconnection.AuthData,
+// and storing a new consumerconnection.Info into the context that it passes down. The
+// attribute names should be documented with their return types and considered
+// part of the public API for the authenticator.
+//
+// # Consumers
+//
+// Provided that the pipeline does not contain processors that would discard or
+// rewrite the context, such as the batch processor, processors and exporters
+// have access to the consumerconnection.Info via consumerconnection.FromContext. Among other usages,
+// this data can be used to:
+//
+// - annotate data points with authentication data (username, tenant, ...)
+//
+// - route data points based on authentication data
+//
+// - rate limit client calls based on IP addresses
+//
+// Processors and exporters relying on the existence of data from the
+// consumerconnection.Info, especially consumerconnection.AuthData, should clearly document this as part
+// of the component's README file. The expected pattern for consuming data is to
+// allow users to specify the attribute name to use in the component. The
+// expected data type should also be communicated to users, who should then
+// compare this with the authenticators that are part of the pipeline. For
+// example, assuming that the OIDC authenticator pushes a "subject" string
+// attribute and that we have a hypothetical "authprinter" processor that prints
+// the "username" to the console, this is how an OpenTelemetry Collector
+// configuration would look like:
+//
+//	extensions:
+//	  oidc:
+//	    issuer_url: http://localhost:8080/auth/realms/opentelemetry
+//	    audience: collector
+//	receivers:
+//	  otlp:
+//	    protocols:
+//	      grpc:
+//	        auth:
+//	          authenticator: oidc
+//	processors:
+//	  authprinter:
+//	    attribute: subject
+//	exporters:
+//	  debug:
+//	service:
+//	  extensions: [oidc]
+//	  pipelines:
+//	    traces:
+//	      receivers: [otlp]
+//	      processors: [authprinter]
+//	      exporters: [debug]
+package consumerconnection // import "go.opentelemetry.io/collector/consumer/consumerconnection"
```

Comment:
`consumerconnection` go likes shorter names, `conn` is a good replacement for `connection`. 

Not sure if is not cleaner to have `consumer.ConnInfo` instead? OR `consumerinfo.Conn` and `consumerinfo.AuthData`?

Reply:
Updated to `consumerconn` for now.  That gives us `consumerconn.Info` and `consumerconn.AuthData` which I like since sharing connection data is the overall theme of the package and `Info` and `AuthData` are pieces of connection data the package supports sharing.